crates/router/src/core/payments/tokenization.rs

```diff
@@ -29,38 +29,62 @@ use crate::{
     utils::{generate_id, OptionExt},
 };
 
+pub struct SavePaymentMethodData<Req> {
+    request: Req,
+    response: Result<types::PaymentsResponseData, types::ErrorResponse>,
+    payment_method_token: Option<types::PaymentMethodToken>,
+    payment_method: PaymentMethod,
+    attempt_status: common_enums::AttemptStatus,
+}
+
+impl<F, Req: Clone> From<&types::RouterData<F, Req, types::PaymentsResponseData>>
+    for SavePaymentMethodData<Req>
+{
+    fn from(router_data: &types::RouterData<F, Req, types::PaymentsResponseData>) -> Self {
+        Self {
+            request: router_data.request.clone(),
+            response: router_data.response.clone(),
+            payment_method_token: router_data.payment_method_token.clone(),
+            payment_method: router_data.payment_method,
+            attempt_status: router_data.status,
+        }
+    }
+}
+
 #[instrument(skip_all)]
 #[allow(clippy::too_many_arguments)]
-pub async fn save_payment_method<F: Clone, FData>(
+pub async fn save_payment_method<FData>(
     state: &AppState,
-    connector: &api::ConnectorData,
-    resp: types::RouterData<F, FData, types::PaymentsResponseData>,
-    maybe_customer: &Option<domain::Customer>,
+    connector_name: String,
+    merchant_connector_id: Option<String>,
+    save_payment_method_data: SavePaymentMethodData<FData>,
+    customer_id: String,
```

Comment:
`customer_id` may not be available here as customer might not be available in all the flows